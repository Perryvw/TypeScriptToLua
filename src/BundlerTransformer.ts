import { TranspiledFile } from "./Transpile";
import { LuaProjectTransformer } from "./CustomLuaTransformers";
import { createBlock, Statement } from "./LuaAST";

type BundlerTransformerFactory = (outFile: string) => LuaProjectTransformer;

function notUndefined<T>(value: T | undefined): value is T {
    return value !== undefined;
}

// TODO: Implement correctly, this is just a PoC!
export const bundlerTransformer: BundlerTransformerFactory = (outFile: string) => (projectFiles: TranspiledFile[]) => {
    const bundledFile: TranspiledFile = {
        fileName: outFile,
        declaration: projectFiles
            .map(f => f.declaration)
            .filter(notUndefined)
            .join("\n"),
        luaAst: createBlock(
            ([] as Statement[]).concat(
                ...projectFiles
                    .map(f => f.luaAst)
                    .filter(notUndefined)
                    .map(ast => ast.statements)
            )
        ),

    };
    return [bundledFile];
};
