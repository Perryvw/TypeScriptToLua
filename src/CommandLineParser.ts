import * as path from "path";
import * as ts from "typescript";
import { CompilerOptions, LuaLibImportKind, LuaTarget } from "./CompilerOptions";
import * as diagnostics from "./diagnostics";

export interface ParsedCommandLine extends ts.ParsedCommandLine {
    options: CompilerOptions;
}

interface CommandLineOptionBase {
    describe: string;
    aliases?: string[];
}

interface CommandLineOptionOfEnum extends CommandLineOptionBase {
    type: "enum";
    choices: string[];
}

interface CommandLineOptionOfBoolean extends CommandLineOptionBase {
    type: "boolean";
}

type CommandLineOption = CommandLineOptionOfEnum | CommandLineOptionOfBoolean;
const optionDeclarations: Record<string, CommandLineOption> = {
    luaLibImport: {
        describe: "Specifies how js standard features missing in lua are imported.",
        type: "enum",
        choices: Object.values(LuaLibImportKind),
    },
    luaTarget: {
        aliases: ["lt"],
        describe: "Specify Lua target version.",
        type: "enum",
        choices: Object.values(LuaTarget),
    },
    noHeader: {
        describe: "Specify if a header will be added to compiled files.",
        type: "boolean",
    },
    noHoisting: {
        describe: "Disables hoisting.",
        type: "boolean",
    },
    sourceMapTraceback: {
        describe: "Applies the source map to show source TS files and lines in error tracebacks.",
        type: "boolean",
    },
};

export const version = `Version ${require("../package.json").version}`;

const helpString =
    "Syntax:   tstl [options] [files...]\n\n" +

    "Examples: tstl path/to/file.ts [...]\n" +
    "          tstl -p path/to/tsconfig.json\n\n" +

    "In addition to the options listed below you can also pass options\n" +
    "for the typescript compiler (For a list of options use tsc -h).\n" +
    "Some tsc options might have no effect.";

export function getHelpString(): string {
    let result = helpString + "\n\n";

    result += "Options:\n";
    for (const [optionName, option] of Object.entries(optionDeclarations)) {
        const aliasStrings = (option.aliases || []).map(a => "-" + a);

        const optionString = aliasStrings.concat(["--" + optionName]).join("|");

        const parameterDescribe = option.type === "enum" ? option.choices.join("|") : option.type;

        const spacing = " ".repeat(Math.max(1, 45 - optionString.length - parameterDescribe.length));

        result += `\n ${optionString} <${parameterDescribe}>${spacing}${option.describe}\n`;
    }

    return result;
}

export function updateParsedConfigFile(parsedConfigFile: ts.ParsedCommandLine): ParsedCommandLine {
    for (const key in parsedConfigFile.raw) {
        const option = optionDeclarations[key];
        if (!option) continue;

        const value = readValue(parsedConfigFile.raw[key], option.type);
        if (option.type === "enum" && !option.choices.includes(value as string)) {
            parsedConfigFile.errors.push(
                diagnostics.argumentForOptionMustBe(key, option.choices.join(", "))
            );
        } else {
            // console.warn(`[Deprectated] TSTL options are moving to the luaConfig object. Adjust your tsconfig to `
            //    + `look like { "compilerOptions": { <typescript options> }, "tstl": { <tstl options> } }`);
            if (parsedConfigFile.options[key] === undefined) {
                parsedConfigFile.options[key] = value;
            }
        }
    }

    // Eventually we will only look for the tstl object for tstl options
    if (parsedConfigFile.raw.tstl) {
        for (const key in parsedConfigFile.raw.tstl) {
            const option = optionDeclarations[key];
            if (!option) continue;

            const value = readValue(parsedConfigFile.raw.tstl[key], option.type);
            if (option.type === "enum" && !option.choices.includes(value as string)) {
                parsedConfigFile.errors.push(
                    diagnostics.argumentForOptionMustBe(key, option.choices.join(", "))
                );
            } else {
                if (parsedConfigFile.options[key] === undefined) {
                    parsedConfigFile.options[key] = value;
                }
            }
        }
    }

    return parsedConfigFile;
}

export function parseCommandLine(args: string[]): ParsedCommandLine {
    const commandLine = updateParsedCommandLine(ts.parseCommandLine(args), args);

    // TODO: Remove
    if (commandLine.options.project && !commandLine.options.rootDir) {
        commandLine.options.rootDir = path.dirname(commandLine.options.project);
    }

    if (!commandLine.options.rootDir) {
        commandLine.options.rootDir = process.cwd();
    }

    if (!commandLine.options.outDir) {
        commandLine.options.outDir = commandLine.options.rootDir;
    }

    return commandLine;
}

function updateParsedCommandLine(
    parsedCommandLine: ts.ParsedCommandLine,
    args: string[]
): ParsedCommandLine {
    // Generate a list of valid option names and aliases
    const optionNames = Object.keys(optionDeclarations)
        .map(n => `--${n}`)
        .concat(...Object.values(optionDeclarations).map(o => (o.aliases || []).map(a => `-${a}`)));

    // Ignore errors caused by tstl specific compiler options
    const tsInvalidCompilerOptionErrorCode = 5023;
    parsedCommandLine.errors = parsedCommandLine.errors.filter(err => {
        return !(
            err.code === tsInvalidCompilerOptionErrorCode &&
            optionNames.some(optionName => String(err.messageText).endsWith(`'${optionName}'.`))
        );
    });

    for (let i = 0; i < args.length; i++) {
        if (!args[i].startsWith("-")) continue;

        const hasTwoDashes = args[i].startsWith("--");
        const parameterValue = args[i].substr(hasTwoDashes ? 2 : 1);
        let argumentName = optionDeclarations[parameterValue] && parameterValue;
        if (!hasTwoDashes && argumentName === undefined) {
            for (const key in optionDeclarations) {
                if ((optionDeclarations[key].aliases || []).includes(parameterValue)) {
                    argumentName = key;
                    break;
                }
            }
        }

        if (argumentName !== undefined) {
            const argumentResult = getArgumentValue(argumentName, i, args);
            if (argumentResult.isValid === true) {
                parsedCommandLine.options[argumentName] = argumentResult.result;
                // Skip value from being considered as option
                i += argumentResult.increment;
            } else {
                parsedCommandLine.errors.push(argumentResult.error);
            }
        }
    }

    return parsedCommandLine;
}

type ArgumentParseResult =
    | { isValid: true; result: string | boolean; increment: number }
    | { isValid: false; error: ts.Diagnostic };

function getArgumentValue(
    argumentName: string,
    argumentIndex: number,
    args: string[]
): ArgumentParseResult {
    const option = optionDeclarations[argumentName];
    const argument = args[argumentIndex + 1];

    if (option.type === "boolean" && (argument === undefined || argument.startsWith("-"))) {
        // Set boolean arguments without supplied value to true
        return { isValid: true, result: true, increment: 0 };
    }

    if (argument === undefined) {
        return { isValid: false, error: diagnostics.compilerOptionExpectsAnArgument(argumentName) };
    }

    const value = readValue(argument, option.type);

    if (option.type === "enum" && option.choices && !option.choices.includes(value as string)) {
        return {
            isValid: false,
            error: diagnostics.argumentForOptionMustBe(
                `--${argumentName}`,
                option.choices.join(", ")
            ),
        };
    }

    return { isValid: true, result: value, increment: 1 };
}

function readValue(value: string | boolean, type: CommandLineOption["type"]): string | boolean {
    if (type === "boolean") {
        return value === true || value === "true" || value === "t";
    } else if (type === "enum") {
        return value.toString().toLowerCase();
    }
}

export function parseConfigFileWithSystem(
    configFileName: string,
    commandLineOptions: CompilerOptions,
    system = ts.sys
): ParsedCommandLine {
    const { config, error } = ts.readConfigFile(configFileName, system.readFile);
    if (error) return { options: {}, fileNames: [], errors: [error] };

    const parsedConfigFile = ts.parseJsonConfigFileContent(
        config,
        system,
        path.dirname(configFileName),
        commandLineOptions,
        configFileName
    );

    return updateParsedConfigFile(parsedConfigFile);
}
