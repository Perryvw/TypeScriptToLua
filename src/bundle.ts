import * as path from "path";
import { SourceNode } from "source-map";
import * as ts from "typescript";
import { couldNotFindBundleEntryPoint } from "./diagnostics";
import { EmitHost, TranspiledFile } from "./Transpile";
import { formatPathToLuaPath, trimExtension } from "./utils";
import { escapeString } from "./TSHelper";
import { CompilerOptions } from "./CompilerOptions";

const formatPath = (path: string) => escapeString(formatPathToLuaPath(trimExtension(path)));
const modulePath = (baseDir: string, pathToResolve: string) => formatPath(path.relative(baseDir, pathToResolve));

export function bundleTranspiledFiles(
    bundleFile: string,
    entryModule: string,
    transpiledFiles: TranspiledFile[],
    program: ts.Program,
    emitHost: EmitHost
): [ts.Diagnostic[], TranspiledFile] {
    const diagnostics: ts.Diagnostic[] = [];

    const options = program.getCompilerOptions() as CompilerOptions;

    const projectRootDir = options.configFilePath
        ? path.dirname(options.configFilePath)
        : emitHost.getCurrentDirectory();

    // Resolve project settings relative to project file.
    const resolvedEntryModule = path.resolve(projectRootDir, entryModule);
    const resolvedBundleFile = path.resolve(projectRootDir, bundleFile);

    // Resolve source files relative to common source directory.
    const sourceRootDir = program.getCommonSourceDirectory();
    if (!transpiledFiles.some(f => path.resolve(sourceRootDir, f.fileName) === resolvedEntryModule)) {
        return [[couldNotFindBundleEntryPoint(entryModule)], { fileName: bundleFile }];
    }

    // For each file: ["<module path>"] = function() <lua content> end,
    const moduleTableEntries: SourceChunk[] = transpiledFiles.map(f =>
        moduleSourceNode(f, modulePath(sourceRootDir, f.fileName))
    );

    // If any of the modules contains a require for lualib_bundle, add it to the module table.
    const lualibRequired = transpiledFiles.some(f => f.lua && f.lua.includes(`require("lualib_bundle")`));
    if (lualibRequired) {
        const lualibBundle = emitHost.readFile(path.resolve(__dirname, "../dist/lualib/lualib_bundle.lua"));
        moduleTableEntries.push(`["lualib_bundle"] = function() ${lualibBundle} end,\n`);
    }

    // Create ____modules table containing all entries from moduleTableEntries
    const moduleTable = createModuleTableNode(moduleTableEntries);

    // Override `require` to read from ____modules table.
    const requireOverride = `
local ____moduleCache = {}
local ____originalRequire = require
function require(file)
    if ____moduleCache[file] then return ____moduleCache[file] end
    if ____modules[file] then
        ____moduleCache[file] = ____modules[file]()
        return ____moduleCache[file]
    else
        print("Could not find module '"..file.."' to require.")
        return ____originalRequire(file) 
    end
end\n`;

    // return require("<entry module path>")
    const entryPoint = `return require("${modulePath(sourceRootDir, resolvedEntryModule)}")\n`;

    const bundleNode = joinSourceChunks([moduleTable, requireOverride, entryPoint]);
    const { code, map } = bundleNode.toStringWithSourceMap();

    return [
        diagnostics,
        {
            fileName: resolvedBundleFile,
            lua: code,
            sourceMap: map.toString(),
            sourceMapNode: moduleTable,
            declaration: undefined,
            declarationMap: undefined,
        },
    ];
}

function moduleSourceNode(transpiledFile: TranspiledFile, modulePath: string): SourceNode {
    const tableEntryHead = `["${modulePath}"] = function() `;
    const tableEntryTail = `end,\n`;

    if (transpiledFile.lua && transpiledFile.sourceMapNode) {
        return joinSourceChunks([tableEntryHead, transpiledFile.sourceMapNode, tableEntryTail]);
    } else {
        return joinSourceChunks([tableEntryHead, tableEntryTail]);
    }
}

function createModuleTableNode(fileChunks: SourceChunk[]): SourceNode {
    const tableHead = `local ____modules = {\n`;
    const tableEnd = `}\n`;

    return joinSourceChunks([tableHead, ...fileChunks, tableEnd]);
}

type SourceChunk = string | SourceNode;
function joinSourceChunks(chunks: SourceChunk[]): SourceNode {
    // tslint:disable-next-line:no-null-keyword
    return new SourceNode(null, null, null, chunks);
}
