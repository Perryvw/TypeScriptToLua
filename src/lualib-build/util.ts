import * as tstl from "..";

export function isExportTableDeclaration(node: tstl.Node): node is tstl.VariableDeclarationStatement & { left: [] } {
    return tstl.isVariableDeclarationStatement(node) && isExportTable(node.left[0]);
}

export function isExportTable(node: tstl.Node): node is tstl.Identifier {
    return tstl.isIdentifier(node) && node.text === "____exports";
}

export type ExportTableIndex = tstl.TableIndexExpression & { index: tstl.StringLiteral };
export function isExportTableIndex(node: tstl.Node): node is ExportTableIndex {
    return tstl.isTableIndexExpression(node) && isExportTable(node.table) && tstl.isStringLiteral(node.index);
}

export function isExportAlias(
    node: tstl.Node,
): node is tstl.VariableDeclarationStatement & { right: [ExportTableIndex] } {
    return tstl.isVariableDeclarationStatement(node) && node.right !== undefined && isExportTableIndex(node.right[0]);
}

export type ExportAssignment = tstl.AssignmentStatement & { left: [ExportTableIndex] };
export function isExportAssignment(node: tstl.Node): node is ExportAssignment {
    return tstl.isAssignmentStatement(node) && isExportTableIndex(node.left[0]);
}

export function isRequire(node: tstl.Node) {
    return (
        tstl.isVariableDeclarationStatement(node) &&
        node.right &&
        tstl.isCallExpression(node.right[0]) &&
        tstl.isIdentifier(node.right[0].expression) &&
        node.right[0].expression.text === "require"
    );
}

export function isImport(node: tstl.Node, importNames: Set<string>) {
    return tstl.isVariableDeclarationStatement(node) && importNames.has(node.left[0].text);
}

export function isExportsReturn(node: tstl.Node) {
    return (
        tstl.isReturnStatement(node) &&
        tstl.isIdentifier(node.expressions[0]) &&
        node.expressions[0].text === "____exports"
    );
}
