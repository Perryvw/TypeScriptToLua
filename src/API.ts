import * as fs from "fs";
import * as path from "path";
import * as ts from "typescript";
import { parseConfigFileContent } from "./CommandLineParser";
import { CompilerOptions } from "./CompilerOptions";
import { getTranspileOutput, TranspiledFile, TranspilationResult } from "./Transpile";

export function transpileFiles(
    rootNames: string[],
    options: CompilerOptions = {}
): TranspilationResult {
    const program = ts.createProgram(rootNames, options);
    const { diagnostics, transpiledFiles } = getTranspileOutput({ program, options });

    const allDiagnostics = ts.sortAndDeduplicateDiagnostics([
        ...ts.getPreEmitDiagnostics(program),
        ...diagnostics,
    ]);

    return { transpiledFiles, diagnostics: [...allDiagnostics] };
}

export function transpileProject(
    fileName: string,
    options?: CompilerOptions
): TranspilationResult {
    const parseResult = parseConfigFileContent(
        fs.readFileSync(fileName, "utf8"),
        fileName,
        options
    );
    if (parseResult.isValid === false) {
        // TODO: Return diagnostics
        throw new Error(parseResult.errorMessage);
    }

    return transpileFiles(parseResult.result.fileNames, parseResult.result.options);
}

const libCache: { [key: string]: ts.SourceFile } = {};
export function createVirtualProgram(
    input: Record<string, string>,
    options?: CompilerOptions
): ts.Program {
    const compilerHost: ts.CompilerHost = {
        fileExists: () => true,
        getCanonicalFileName: fileName => fileName,
        getCurrentDirectory: () => "",
        getDefaultLibFileName: ts.getDefaultLibFileName,
        readFile: () => "",
        getNewLine: () => "\n",
        useCaseSensitiveFileNames: () => false,
        writeFile: () => {},

        getSourceFile: filename => {
            if (filename in input) {
                return ts.createSourceFile(
                    filename,
                    input[filename],
                    ts.ScriptTarget.Latest,
                    false
                );
            }

            if (filename.startsWith("lib.")) {
                if (libCache[filename]) return libCache[filename];
                const typeScriptDir = path.dirname(require.resolve("typescript"));
                const filePath = path.join(typeScriptDir, filename);
                const content = fs.readFileSync(filePath, "utf8");

                libCache[filename] = ts.createSourceFile(
                    filename,
                    content,
                    ts.ScriptTarget.Latest,
                    false
                );

                return libCache[filename];
            }
        },
    };

    return ts.createProgram(Object.keys(input), options, compilerHost);
}

export interface TranspileStringResult {
    file: TranspiledFile;
    diagnostics: ts.Diagnostic[];
}

export function transpileString(
    input: string | Record<string, string>,
    options: CompilerOptions = {}
): TranspileStringResult {
    const programFiles = typeof input === "object" ? input : { "main.ts": input };
    const mainFileName =
        typeof input === "string"
            ? "main.ts"
            : Object.keys(input).find(x => /\bmain\.[a-z]+$/.test(x));
    if (mainFileName === undefined) throw new Error('Input should have a file named "main"');

    const program = createVirtualProgram(programFiles, options);
    const { diagnostics, transpiledFiles } = getTranspileOutput({ program, options });

    const allDiagnostics = ts.sortAndDeduplicateDiagnostics([
        ...ts.getPreEmitDiagnostics(program),
        ...diagnostics,
    ]);

    return { file: transpiledFiles.get(mainFileName), diagnostics: [...allDiagnostics] };
}
