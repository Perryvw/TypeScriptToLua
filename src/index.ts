import * as fs from "fs";
import * as path from "path";
import * as ts from "typescript";
import { parseConfigFileWithSystem } from "./CommandLineParser";
import { CompilerOptions } from "./CompilerOptions";
import { getTranspileOutput, TranspilationResult, TranspiledFile } from "./Transpile";

export { parseCommandLine, ParsedCommandLine, updateParsedConfigFile } from "./CommandLineParser";
export { CompilerOptions, LuaLibImportKind, LuaTarget } from "./CompilerOptions";
export * from "./Emit";
export * from "./LuaAST";
export { LuaLibFeature } from "./LuaLib";
export { LuaPrinter } from "./LuaPrinter";
export { LuaTransformer } from "./LuaTransformer";
export * from "./Transpile";

export function transpileFiles(
    rootNames: string[],
    options: CompilerOptions = {}
): TranspilationResult {
    const program = ts.createProgram(rootNames, options);
    const { diagnostics, transpiledFiles } = getTranspileOutput({ program, options });

    const allDiagnostics = ts.sortAndDeduplicateDiagnostics([
        ...ts.getPreEmitDiagnostics(program),
        ...diagnostics,
    ]);

    return { transpiledFiles, diagnostics: [...allDiagnostics] };
}

export function transpileProject(fileName: string, options?: CompilerOptions): TranspilationResult {
    const parseResult = parseConfigFileWithSystem(fileName, options);
    if (parseResult.errors.length > 0) {
        return { diagnostics: parseResult.errors, transpiledFiles: new Map() };
    }

    return transpileFiles(parseResult.fileNames, parseResult.options);
}

const libCache: { [key: string]: ts.SourceFile } = {};

/** @internal */
export function createVirtualProgram(
    input: Record<string, string>,
    options?: CompilerOptions
): ts.Program {
    const compilerHost: ts.CompilerHost = {
        fileExists: () => true,
        getCanonicalFileName: fileName => fileName,
        getCurrentDirectory: () => "",
        getDefaultLibFileName: ts.getDefaultLibFileName,
        readFile: () => "",
        getNewLine: () => "\n",
        useCaseSensitiveFileNames: () => false,
        writeFile: () => {},

        getSourceFile: filename => {
            if (filename in input) {
                return ts.createSourceFile(
                    filename,
                    input[filename],
                    ts.ScriptTarget.Latest,
                    false
                );
            }

            if (filename.startsWith("lib.")) {
                if (libCache[filename]) return libCache[filename];
                const typeScriptDir = path.dirname(require.resolve("typescript"));
                const filePath = path.join(typeScriptDir, filename);
                const content = fs.readFileSync(filePath, "utf8");

                libCache[filename] = ts.createSourceFile(
                    filename,
                    content,
                    ts.ScriptTarget.Latest,
                    false
                );

                return libCache[filename];
            }
        },
    };

    return ts.createProgram(Object.keys(input), options, compilerHost);
}

export interface VirtualProgramResult {
    file: TranspiledFile;
    diagnostics: ts.Diagnostic[];
}

export function transpileString(
    content: string,
    options: CompilerOptions = {}
): VirtualProgramResult {
    const program = createVirtualProgram({ "main.ts": content }, options);
    const { diagnostics, transpiledFiles } = getTranspileOutput({ program, options });
    const allDiagnostics = ts.sortAndDeduplicateDiagnostics([
        ...ts.getPreEmitDiagnostics(program),
        ...diagnostics,
    ]);

    return { file: transpiledFiles.get("main.ts"), diagnostics: [...allDiagnostics] };
}

export function transpileVirtualProgram(
    files: Record<string, string>,
    options: CompilerOptions = {}
): VirtualProgramResult {
    const mainFileName = Object.keys(files).find(x => /\bmain\.[a-z]+$/.test(x));
    if (mainFileName === undefined) {
        throw new Error('Virtual program should have a file named "main"');
    }

    const program = createVirtualProgram(files, options);
    const { diagnostics, transpiledFiles } = getTranspileOutput({ program, options });
    const allDiagnostics = ts.sortAndDeduplicateDiagnostics([
        ...ts.getPreEmitDiagnostics(program),
        ...diagnostics,
    ]);

    return { file: transpiledFiles.get(mainFileName), diagnostics: [...allDiagnostics] };
}
