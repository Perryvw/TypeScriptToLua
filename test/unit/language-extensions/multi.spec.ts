import * as path from "path";
import * as util from "../../util";
import * as tstl from "../../../src";
import {
    invalidMultiFunctionUse,
    invalidMultiReturnAccess,
    invalidMultiFunctionReturnType,
} from "../../../src/transformation/utils/diagnostics";

const multiProjectOptions: tstl.CompilerOptions = {
    types: [path.resolve(__dirname, "../../../language-extensions")],
};

test("multi example use case", () => {
    util.testModule`
        function multiReturn(): LuaMultiReturn<[string, number]> {
            return $multi("foo", 5);
        }

        const [a, b] = multiReturn();
        export { a, b };
    `
        .setOptions(multiProjectOptions)
        .expectToEqual({ a: "foo", b: 5 });
});

test.each<[string, any]>([
    ["$multi()", undefined],
    ["$multi(true)", true],
    ["$multi(1, 2)", 1],
])("$multi call on return statement (%s)", (expression, result) => {
    util.testFunction`
        return ${expression};
    `
        .setOptions(multiProjectOptions)
        .expectToEqual(result);
});

const multiFunction = `
function multi(...args) {
    return $multi(...args);
}
`;

const createCasesThatCall = (name: string): Array<[string, any]> => [
    [`let a; [a] = ${name}()`, undefined],
    [`const [a] = ${name}()`, undefined],
    [`const [a] = ${name}(1)`, 1],
    [`const [a = 1] = ${name}()`, 1],
    [`const [a = 1] = ${name}(2)`, 2],
    [`const ar = [1]; const [a] = ${name}(...ar)`, 1],
    [`const _ = null, [a] = ${name}(1)`, 1],
    [`let a; for (const [a] = ${name}(1, 2); false; 1) {}`, undefined],
    [`let a; for ([a] = ${name}(1, 2); false; 1) {}`, 1],
    [`let a; if ([a] = ${name}(1)) { ++a; }`, 2],
];

test.each<[string, any]>(createCasesThatCall("$multi"))("invalid direct $multi function use (%s)", statement => {
    util.testModule`
        ${multiFunction}
        ${statement}
        export { a };
    `
        .setOptions(multiProjectOptions)
        .setReturnExport("a")
        .expectDiagnosticsToMatchSnapshot([invalidMultiFunctionUse.code]);
});

test.each<[string, any]>(createCasesThatCall("multi"))(
    "valid indirect $multi function use (%s)",
    (statement, result) => {
        util.testModule`
            ${multiFunction}
            ${statement}
            export { a };
        `
            .setOptions(multiProjectOptions)
            .setReturnExport("a")
            .expectToEqual(result);
    }
);

test.each<[string, number[]]>([
    ["$multi", [invalidMultiFunctionUse.code]],
    ["$multi()", [invalidMultiFunctionUse.code]],
    ["({ $multi });", [invalidMultiFunctionUse.code]],
    ["const a = $multi();", [invalidMultiFunctionUse.code]],
    ["const {} = $multi();", [invalidMultiFunctionUse.code]],
    ["([a] = $multi(1)) => {}", [invalidMultiFunctionUse.code]],
    ["const [a = 0] = $multi()", [invalidMultiFunctionUse.code]],
])("invalid $multi call (%s)", (statement, diagnostics) => {
    util.testModule`
        ${statement}
    `
        .setOptions(multiProjectOptions)
        .expectDiagnosticsToMatchSnapshot(diagnostics);
});

test("function to spread multi type result from multi type function", () => {
    util.testFunction`
        ${multiFunction}
        function m() {
            return $multi(...multi(true));
        }
        return m();
    `
        .setOptions(multiProjectOptions)
        .expectToEqual(true);
});

test("$multi call with destructuring assignment side effects", () => {
    util.testModule`
        ${multiFunction}
        let a;
        export { a };
        [a] = multi(1);
    `
        .setOptions(multiProjectOptions)
        .setReturnExport("a")
        .expectToEqual(1);
});

test("allow $multi call in ArrowFunction body", () => {
    util.testFunction`
        const call = () => $multi(1);
        const [result] = call();
        return result;
    `
        .setOptions(multiProjectOptions)
        .expectToEqual(1);
});

test("forward $multi call in ArrowFunction body", () => {
    util.testFunction`
        const foo = () => $multi(1);
        const call = () => foo();
        const [result] = call();
        return result;
    `
        .setOptions(multiProjectOptions)
        .expectToEqual(1);
});

test.each(["0", "i"])("allow LuaMultiReturn numeric access (%s)", expression => {
    util.testFunction`
        ${multiFunction}
        const i = 0;
        return multi(1)[${expression}];
    `
        .setOptions(multiProjectOptions)
        .expectToEqual(1);
});

test.each(["multi()['forEach']", "multi().forEach"])("disallow LuaMultiReturn non-numeric access", expression => {
    util.testFunction`
        ${multiFunction}
        return ${expression};
    `
        .setOptions(multiProjectOptions)
        .expectDiagnosticsToMatchSnapshot([invalidMultiReturnAccess.code]);
});

test("invalid $multi implicit cast", () => {
    util.testModule`
        function badMulti(): [string, number] {
            return $multi("foo", 42);
        }
    `
        .setOptions(multiProjectOptions)
        .expectDiagnosticsToMatchSnapshot([invalidMultiFunctionReturnType.code]);
});

test.each([
    { flag: true, result: 4 },
    { flag: false, result: 7 },
])("overloaded $multi/non-$multi return", ({ flag, result }) => {
    util.testFunction`
        function multiOverload(flag: true): LuaMultiReturn<[string, number]>;
        function multiOverload(flag: false): [string, number];
        function multiOverload(flag: boolean) {
            if (flag) {
                return $multi("foo", 4);
            } else {
                return ["bar", 7];
            }
        }
        const [x, y] = multiOverload(${flag});
        return y;
    `
        .setOptions(multiProjectOptions)
        .expectToEqual(result);
});

test("return $multi from try", () => {
    util.testFunction`
        function multiTest() {
            try {
                return $multi(1, 2);
            } catch {
            }
        }
        const [_, a] = multiTest();
        return a;
    `
        .setOptions(multiProjectOptions)
        .expectToEqual(2);
});

test("return $multi from catch", () => {
    util.testFunction`
        function multiTest() {
            try {
                throw "error";
            } catch {
                return $multi(1, 2);
            }
        }
        const [_, a] = multiTest();
        return a;
    `
        .setOptions(multiProjectOptions)
        .expectToEqual(2);
});

test("return LuaMultiReturn from try", () => {
    util.testFunction`
        ${multiFunction}
        function multiTest() {
            try {
                return multi(1, 2);
            } catch {
            }
        }
        const [_, a] = multiTest();
        return a;
    `
        .setOptions(multiProjectOptions)
        .expectToEqual(2);
});

test("return LuaMultiReturn from catch", () => {
    util.testFunction`
        ${multiFunction}
        function multiTest() {
            try {
                throw "error";
            } catch {
                return multi(1, 2);
            }
        }
        const [_, a] = multiTest();
        return a;
    `
        .setOptions(multiProjectOptions)
        .expectToEqual(2);
});
