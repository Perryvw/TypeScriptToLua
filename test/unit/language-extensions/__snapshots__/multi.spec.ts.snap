// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`invalid $multi call ($multi()): code 1`] = `"____(_G)"`;

exports[`invalid $multi call ($multi()): diagnostics 1`] = `"main.ts(2,9): error TSTL: The $multi function must be called in return statements."`;

exports[`invalid $multi call ($multi): code 1`] = `"local ____ = ____"`;

exports[`invalid $multi call ($multi): diagnostics 1`] = `"main.ts(2,9): error TSTL: The $multi function must be called in return statements."`;

exports[`invalid $multi call (([a] = $multi(1)) => {}): code 1`] = `
"local function ____(____, ____bindingPattern0)
    if ____bindingPattern0 == nil then
        ____bindingPattern0 = ____(_G, 1)
    end
    local a = ____bindingPattern0[1]
end"
`;

exports[`invalid $multi call (([a] = $multi(1)) => {}): diagnostics 1`] = `"main.ts(2,16): error TSTL: The $multi function must be called in return statements."`;

exports[`invalid $multi call (({ $multi });): code 1`] = `"local ____ = nil"`;

exports[`invalid $multi call (({ $multi });): diagnostics 1`] = `"main.ts(2,12): error TSTL: The $multi function must be called in return statements."`;

exports[`invalid $multi call (const [a = 0] = $multi()): code 1`] = `""`;

exports[`invalid $multi call (const [a = 0] = $multi()): diagnostics 1`] = `"main.ts(2,15): error TSTL: This array binding pattern cannot have initializers."`;

exports[`invalid $multi call (const {} = $multi();): code 1`] = `""`;

exports[`invalid $multi call (const {} = $multi();): diagnostics 1`] = `"main.ts(2,15): error TSTL: Expected an array destructuring pattern."`;

exports[`invalid $multi call (const a = $multi();): code 1`] = `""`;

exports[`invalid $multi call (const a = $multi();): diagnostics 1`] = `"main.ts(2,15): error TSTL: Expected an array destructuring pattern."`;

exports[`invalid direct $multi function use (const [a] = $multi()): code 1`] = `
"local ____exports = {}
local function multi(self, ...)
    local args = {...}
    return table.unpack(args)
end
____exports.a = a
return ____exports"
`;

exports[`invalid direct $multi function use (const [a] = $multi()): diagnostics 1`] = `"main.ts(7,21): error TSTL: The $multi function must be called in return statements."`;

exports[`invalid direct $multi function use (const [a] = $multi(1)): code 1`] = `
"local ____exports = {}
local function multi(self, ...)
    local args = {...}
    return table.unpack(args)
end
____exports.a = a
return ____exports"
`;

exports[`invalid direct $multi function use (const [a] = $multi(1)): diagnostics 1`] = `"main.ts(7,21): error TSTL: The $multi function must be called in return statements."`;

exports[`invalid direct $multi function use (const _ = null, [a] = $multi(1)): code 1`] = `
"local ____exports = {}
local function multi(self, ...)
    local args = {...}
    return table.unpack(args)
end
local _ = nil
____exports.a = a
return ____exports"
`;

exports[`invalid direct $multi function use (const _ = null, [a] = $multi(1)): diagnostics 1`] = `"main.ts(7,31): error TSTL: The $multi function must be called in return statements."`;

exports[`invalid direct $multi function use (const ar = [1]; const [a] = $multi(...ar)): code 1`] = `
"local ____exports = {}
local function multi(self, ...)
    local args = {...}
    return table.unpack(args)
end
local ar = {1}
____exports.a = a
return ____exports"
`;

exports[`invalid direct $multi function use (const ar = [1]; const [a] = $multi(...ar)): diagnostics 1`] = `"main.ts(7,37): error TSTL: The $multi function must be called in return statements."`;

exports[`invalid direct $multi function use (let a; [a] = $multi()): code 1`] = `
"local ____exports = {}
local function multi(self, ...)
    local args = {...}
    return table.unpack(args)
end
local a
____exports.a = a
return ____exports"
`;

exports[`invalid direct $multi function use (let a; [a] = $multi()): diagnostics 1`] = `"main.ts(7,22): error TSTL: The $multi function must be called in return statements."`;

exports[`invalid direct $multi function use (let a; for ([a] = $multi(1, 2); false; 1) {}): code 1`] = `
"local ____exports = {}
local function multi(self, ...)
    local args = {...}
    return table.unpack(args)
end
local a
do
    while false do
        local ____ = 1
    end
end
____exports.a = a
return ____exports"
`;

exports[`invalid direct $multi function use (let a; for ([a] = $multi(1, 2); false; 1) {}): diagnostics 1`] = `"main.ts(7,27): error TSTL: The $multi function must be called in return statements."`;

exports[`invalid direct $multi function use (let a; for (const [a] = $multi(1, 2); false; 1) {}): code 1`] = `
"local ____exports = {}
local function multi(self, ...)
    local args = {...}
    return table.unpack(args)
end
local a
do
    while false do
        local ____ = 1
    end
end
____exports.a = a
return ____exports"
`;

exports[`invalid direct $multi function use (let a; for (const [a] = $multi(1, 2); false; 1) {}): diagnostics 1`] = `"main.ts(7,33): error TSTL: The $multi function must be called in return statements."`;
